### PR TITLE
feat: add authentication support

### PR DESCRIPTION
**Pull request by gruyaume**
# Description

Here we add authentication support to the NMS K8s charm. As this is an opt-in feature in the workload, we enable it via the `enableAuthentication` configuration option. On configure event, the charm now creates the first user in NMS, stores the login information in a Juju secret and uses it to make further API call to NMS. The NMS client had to be expanded to include new `login`, `create_first_user`, `is_initialized` methods.

## Dependencies

This PR depends on this upstream change:
- https://github.com/omec-project/webconsole/pull/217

It also depends on us using the new upstream webconsole in the NMS rock:
- https://github.com/canonical/sdcore-nms/pull/659

# Checklist:

- [ ] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that validate the behaviour of the software
- [ ] I validated that new and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have bumped the version of the library

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -408,5 +526,24 @@ def _environment_variables(self) -> dict:
...
+    """Generate a password for the NMS Account."""
+    pw = []
+    pw.append(random.choice(string.ascii_lowercase))
+    pw.append(random.choice(string.ascii_uppercase))
+    pw.append(random.choice(string.digits))
+    pw.append(random.choice(string.punctuation))
+    for i in range(8):
+        pw.append(random.choice(string.ascii_letters + string.digits + string.punctuation))
+    random.shuffle(pw)
+    return "".join(pw)
```
[the remark this answers is not in the document]
Comment (gruyaume) [reply src/charm.py:539]: Haa ok, I mostly copied the code from our Notary charm, we likely will want to make the same changes over there as well.

File: charmcraft.yaml [hunk 2]
```diff
@@ -40,6 +40,8 @@ requires:
     interface: mongodb_client
   auth_database:
     interface: mongodb_client
+  webui_database:
+    interface: mongodb_client
```
[the remark this answers is not in the document]
Comment (gruyaume) [reply charmcraft.yaml:44]: Fixed

File: src/charm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (gruyaume) [reply src/charm.py]: Done

File: src/charm.py [hunk 3]
```diff
@@ -408,5 +526,24 @@ def _environment_variables(self) -> dict:
...
+
+
+def _generate_username() -> str:
+    """Generate a username for the NMS Account."""
+    suffix = [random.choice(string.ascii_uppercase) for i in range(4)]
+    return "charm-admin-" + "".join(suffix)
```
[the remark this answers is not in the document]
Comment (gruyaume) [reply src/charm.py]: Done

**Review by ghislainbourgeois — APPROVED** [review]

**Pull request merged**